src/VisualStudio/Core/Def/Implementation/Workspace/VisualStudioActiveDocumentTracker.cs

```diff
@@ -0,0 +1,290 @@
+                {
+                    return listener.GetDocumentId(workspace);
+                }
+            }
+
+            return null;
+        }
+
+        /// <summary>
+        /// Get a read-only collection of the <see cref="DocumentId"/>s of all the visible documents in the given <see cref="Workspace"/>.
+        /// </summary>
+        public ImmutableArray<DocumentId> GetVisibleDocuments(Workspace workspace)
+        {
+            ThisCanBeCalledOnAnyThread();
+
+            var visibleFramesSnapshot = _visibleFrames;
+
+            var ids = ArrayBuilder<DocumentId>.GetInstance(visibleFramesSnapshot.Count);
+
+            foreach (var frame in visibleFramesSnapshot)
+            {
+                var documentId = frame.GetDocumentId(workspace);
+
+                if (documentId != null)
+                {
+                    ids.Add(documentId);
+                }
+            }
+
+            return ids.ToImmutableAndFree();
+        }
+
+        public void TrackWindowFrame(IVsWindowFrame frame)
+        {
+            AssertIsForeground();
+
+            Contract.ThrowIfNull(frame);
+
+            foreach (var frameListener in _visibleFrames)
+            {
+                if (frameListener.Frame == frame)
+                {
+                    // Already in the visible list
+                    return;
+                }
+            }
+
+            _visibleFrames = _visibleFrames.Add(new FrameListener(this, frame));
+
+            this.DocumentsChanged?.Invoke(this, EventArgs.Empty);
+        }
+
+        private void RemoveFrame(FrameListener frame)
+        {
+            _visibleFrames = _visibleFrames.Remove(frame);
+
+            this.DocumentsChanged?.Invoke(this, EventArgs.Empty);
+        }
+
+        public int OnSelectionChanged(IVsHierarchy pHierOld, [ComAliasName("Microsoft.VisualStudio.Shell.Interop.VSITEMID")]uint itemidOld, IVsMultiItemSelect pMISOld, ISelectionContainer pSCOld, IVsHierarchy pHierNew, [ComAliasName("Microsoft.VisualStudio.Shell.Interop.VSITEMID")]uint itemidNew, IVsMultiItemSelect pMISNew, ISelectionContainer pSCNew)
+        {
+            return VSConstants.E_NOTIMPL;
+        }
+
+        public int OnElementValueChanged([ComAliasName("Microsoft.VisualStudio.Shell.Interop.VSSELELEMID")]uint elementid, object varValueOld, object varValueNew)
+        {
+            AssertIsForeground();
+
+            if (elementid == (uint)VSConstants.VSSELELEMID.SEID_DocumentFrame)
+            {
+                // Remember the newly activated frame so it can be read from another thread.
+                _activeFrame = varValueNew as IVsWindowFrame;
+
+                if (_activeFrame != null)
+                {
+                    // This might be the first time we've ever seen this IVsWindowFrame, so add it to our tracking too
+                    TrackWindowFrame(_activeFrame);
+                }
+
+                this.DocumentsChanged?.Invoke(this, EventArgs.Empty);
```

Comment:
this looks like we'll get a double DocumentChange notification the first time we hear about something.

Reply:
Fixed by refactoring some helpers.